File: medium/Linked List/Insert Greatest Common Divisors in Linked List.cpp
class Solution {
public:
    ListNode* insertGreatestCommonDivisors(ListNode* head) {
        if(head == nullptr || head->next == nullptr) return head;

        ListNode* prev = head;
        ListNode* curr = head->next;

        while(curr != nullptr){
            int x = gcd(prev->val,curr->val);
            ListNode* g = new ListNode(x);

            prev->next = g;
            g->next = curr;

            prev = curr;
            curr = curr->next;
        }return head;
    }
private:
    int gcd(int a,int b){
        while(b!=0){
            int temp = b;
            b = a%b;
            a = temp;
        }return a;
    }
};

File: medium/linked_list/Insert Greatest Common Divisors in Linked List.py
class Solution(object):
    def insertGreatestCommonDivisors(self, head):
        if not head or not head.next:
            return head
        def gcd(a,b):
            while b:
                a,b = b,a%b
            return a

        prev = head
        while prev.next:
            curr = prev.next
            x = gcd(prev.val,curr.val)
            g = ListNode(x)
            prev.next = g
            g.next = curr
            prev = curr
        return head
            
        